Firewall: T4651: Add options to match packet size on firewall rules.

nicolas-fort opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Add options to match packet size on firewall rules.
## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T4651

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
firewall
## Proposed changes
<!--- Describe your changes in detail -->

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
cli:
```
vyos@fwall-ip# set firewall name FOO rule 10 
Possible completions:
   action               Rule action
 > connection-status    Connection status
   description          Description
 > destination          Destination parameters
   disable              Option to disable firewall rule
 > fragment             IP fragment match
 > icmp                 ICMP type and code information
 > ipsec                Inbound IPsec packets
 > limit                Rate limit using a token bucket filter
   log                  Option to log packets matching rule
   log-level            Set log-level. Log must be enable.
   packet-length        Payload size in bytes, including header and data
   protocol             Protocol to match (protocol name, number, or "all")
 > recent               Parameters for matching recently seen sources
 > source               Source parameters
 > state                Session state
 > tcp                  TCP flags to match
 > time                 Time to match rule
 > ttl                  Time to live limit

      
[edit]
vyos@fwall-ip# set firewall name FOO rule 10 packet-length 
Possible completions:
   <1-65535>            Packet length value. Multiple values can be specified as a comma-separated list. Inverted match is also supported
   <start-end>          Packet length range. Inverted match is also supported (e.g. 1001-1005 or !1001-1005)
                        

vyos@fwall-ip# set firewall ipv6-name FOO-6 rule 10 
Possible completions:
   action               Rule action
 > connection-status    Connection status
   description          Description
 > destination          Destination parameters
   disable              Option to disable firewall rule
 > fragment             IP fragment match
 > hop-limit            Hop Limit
 > icmpv6               ICMPv6 type and code information
 > ipsec                Inbound IPsec packets
 > limit                Rate limit using a token bucket filter
   log                  Option to log packets matching rule
   log-level            Set log-level. Log must be enable.
   packet-length        Payload size in bytes, including header and data
   protocol             Protocol to match (protocol name, number, or "all")
 > recent               Parameters for matching recently seen sources
 > source               Source parameters
 > state                Session state
 > tcp                  TCP flags to match
 > time                 Time to match rule


```
Configuration example and validation:
```
set firewall name FOO rule 20 action 'accept'
set firewall name FOO rule 20 packet-length '1500'
## Range match. Ex: allow length < 40000
set firewall name FOO rule 30 action 'accept'
set firewall name FOO rule 30 packet-length '0-40000'
# Inverted match
set firewall name FOO rule 40 action 'drop'
set firewall name FOO rule 40 packet-length '!300'
# Multiple values match
set firewall name FOO rule 50 action 'reject'
set firewall name FOO rule 50 packet-length '2345,3456,8710'
# Inverted range match. Ex: again allow length < 40000
set firewall name FOO rule 60 action 'accept'
set firewall name FOO rule 60 packet-length '!40000-65535'


######## Ipv6 Tests ########
## Single match
set firewall ipv6-name FOO-6 rule 20 action 'accept'
set firewall ipv6-name FOO-6 rule 20 packet-length '1500'
## Range match. Ex: allow length < 40000
set firewall ipv6-name FOO-6 rule 30 action 'accept'
set firewall ipv6-name FOO-6 rule 30 packet-length '0-40000'
# Inverted match
set firewall ipv6-name FOO-6 rule 40 action 'drop'
set firewall ipv6-name FOO-6 rule 40 packet-length '!300'
# Multiple values match
set firewall ipv6-name FOO-6 rule 50 action 'reject'
set firewall ipv6-name FOO-6 rule 50 packet-length '2345,3456,8710'
# Inverted range match. Ex: again allow length < 40000
set firewall ipv6-name FOO-6 rule 60 action 'accept'
set firewall ipv6-name FOO-6 rule 60 packet-length '!40000-65535'

##### Check chains #####
vyos@fwall-ip# sudo nft list chain ip filter NAME_FOO
table ip filter {
        chain NAME_FOO {
                ip length { 1500 } counter packets 0 bytes 0 return comment "FOO-20"
                ip length { 0-40000 } counter packets 0 bytes 0 return comment "FOO-30"
                ip length != { 300 } counter packets 0 bytes 0 drop comment "FOO-40"
                ip length { 2345, 3456, 8710 } counter packets 0 bytes 0 reject comment "FOO-50"
                ip length != { 40000-65535 } counter packets 0 bytes 0 return comment "FOO-60"
                counter packets 0 bytes 0 return comment "FOO default-action accept"
        }
}
[edit]
vyos@fwall-ip# sudo nft list chain ip6 filter NAME6_FOO-6
table ip6 filter {
        chain NAME6_FOO-6 {
                ip6 length { 1500 } counter packets 0 bytes 0 return comment "FOO-6-20"
                ip6 length { 0-40000 } counter packets 0 bytes 0 return comment "FOO-6-30"
                ip6 length != { 300 } counter packets 0 bytes 0 drop comment "FOO-6-40"
                ip6 length { 2345, 3456, 8710 } counter packets 0 bytes 0 reject comment "FOO-6-50"
                ip6 length != { 40000-65535 } counter packets 0 bytes 0 return comment "FOO-6-60"
                counter packets 0 bytes 0 return comment "FOO-6 default-action accept"
        }
}

```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

nicolas-fort committed:
Firewall: T4651: Add options to match packet size on firewall rules.

sever-sever commented on interface-definitions/firewall.xml.in:
```diff
@@ -383,6 +383,26 @@
                 </children>
               </node>
               #include <include/firewall/common-rule.xml.i>
+              <leafNode name="ip-length">
```
As to me it should be “packet-length”

nicolas-fort replied:
Since in ipv4 its "total length" and in ipv6 is "payload length", I chose "ip-length" so it's user-friendly, and same option in both protocols.

I think if changes are necessary, I would go for those values, instead of "packet-length". But of course, open to discussion and more opinions.
@dmbaturin @c-po @zdc

c-po replied:
+1 on `packet-length`, as `ip` is sometimes associated with IPv4.

dmbaturin replied:
Same here. `ip-length` also doesn't really help to resolve the confusion between IPv4 and IPv6 behavior, and most people will associate "IP" with "IPv4", so it will only create more confusion.

c-po commented on interface-definitions/firewall.xml.in:
```diff
@@ -383,6 +383,26 @@
+                  </valueHelp>
+                  <valueHelp>
+                    <format> </format>
+                    <description>\n\n  Multiple values can be specified as a comma-separated list.\n  For example: '64, 512,1001-1005'</description>
```
Please do NOT use empty `<format>` blocks to add additional text. If multiple values are supported, please use `<multi/>`

c-po commented on interface-definitions/firewall.xml.in:
```diff
@@ -571,6 +591,26 @@
                 </children>
               </node>
               #include <include/firewall/common-rule.xml.i>
+              <leafNode name="ip-length">
```
As this node is used twice, please move it to a dedicated `.xml.i` file and then `#include <>` it.

nicolas-fort replied:
It's almost the same, but differs on the  help information, which is slightly different for both protocols.

c-po replied:
You will find a help string that matches for both like we donin anlot of other places. No need to duplicate code for some words…

c-po replied:
`<help>Packet size in bytes, including header and data</help>` will match for both IPv4 and IPv6 as the main information is "including header and data".

nicolas-fort committed:
Firewall: T4651: Change proposed cli from ip-length to packet-length

nicolas-fort commented:
I have changed cli from ip-length to packet-length.
I have also simplified options and helps in the cli.

@c-po I wasn't able to add <multi/> and allow multiple values, including ranges and inverted selections.. My python knowledge is very basic, and after struggling several times, couldn't get it working as expected. So, this is what I was able to do so far, and is working as exposed on the updated task description
